File: python/src/trezorlib/transport/thp/control_byte.py
# This file is part of the Trezor project.
#
# Copyright (C) SatoshiLabs and contributors
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License version 3
# as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the License along with this library.
# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.

from typing import Optional

CODEC_V1 = 0x3F
CONTINUATION_PACKET = 0x80
HANDSHAKE_INIT_REQ = 0x00
HANDSHAKE_INIT_RES = 0x01
HANDSHAKE_COMP_REQ = 0x02
HANDSHAKE_COMP_RES = 0x03
ENCRYPTED_TRANSPORT = 0x04

CONTINUATION_PACKET_MASK = 0x80
ACK_MASK = 0xF7
DATA_MASK = 0xE7

ACK_MESSAGE = 0x20
_ERROR = 0x42
CHANNEL_ALLOCATION_REQ = 0x40
_CHANNEL_ALLOCATION_RES = 0x41

TREZOR_STATE_UNPAIRED = b"\x00"
TREZOR_STATE_PAIRED = b"\x01"


def add_seq_bit_to_ctrl_byte(ctrl_byte: int, seq_bit: int) -> int:
    if seq_bit == 0:
        return ctrl_byte & 0xEF
    if seq_bit == 1:
        return ctrl_byte | 0x10
    raise Exception("Unexpected sequence bit")


def add_ack_bit_to_ctrl_byte(ctrl_byte: int, ack_bit: int) -> int:
    if ack_bit == 0:
        return ctrl_byte & 0xF7
    if ack_bit == 1:
        return ctrl_byte | 0x08
    raise Exception("Unexpected acknowledgement bit")


def get_ack_bit(ctrl_byte: int) -> Optional[int]:
    if ctrl_byte & 0xE0 and not is_ack(ctrl_byte):
        # not all message types contain ACK bit
        return None

    return (ctrl_byte & 0x08) >> 3


def get_seq_bit(ctrl_byte: int) -> Optional[int]:
    if ctrl_byte & 0xE0:
        # not all message types contain SEQ bit
        return None

    return (ctrl_byte & 0x10) >> 4


def is_ack(ctrl_byte: int) -> bool:
    return ctrl_byte & ACK_MASK == ACK_MESSAGE


def is_error(ctrl_byte: int) -> bool:
    return ctrl_byte == _ERROR


def is_continuation(ctrl_byte: int) -> bool:
    return ctrl_byte & CONTINUATION_PACKET_MASK == CONTINUATION_PACKET


def is_encrypted_transport(ctrl_byte: int) -> bool:
    return ctrl_byte & DATA_MASK == ENCRYPTED_TRANSPORT


def is_handshake_init_req(ctrl_byte: int) -> bool:
    return ctrl_byte & DATA_MASK == HANDSHAKE_INIT_REQ


def is_handshake_comp_req(ctrl_byte: int) -> bool:
    return ctrl_byte & DATA_MASK == HANDSHAKE_COMP_REQ

File: rust/trezor-thp/src/control_byte.rs
use crate::{alternating_bit::SyncBits, error::Error, header::HandshakeMessage};

const CODEC_V1: u8 = 0x3F;
const CONTINUATION_PACKET: u8 = 0x80;

const CONTINUATION_PACKET_MASK: u8 = 0x80;

const ACK_MASK: u8 = 0xF7;
const ACK_MESSAGE: u8 = 0x20;
const DATA_MASK: u8 = 0xE7;
const HANDSHAKE_INIT_REQ: u8 = 0x00;
const HANDSHAKE_INIT_RES: u8 = 0x01;
const HANDSHAKE_COMP_REQ: u8 = 0x02;
const HANDSHAKE_COMP_RES: u8 = 0x03;
const ENCRYPTED_TRANSPORT: u8 = 0x04;

const CHANNEL_ALLOCATION_REQ: u8 = 0x40;
const CHANNEL_ALLOCATION_RES: u8 = 0x41;
const ERROR: u8 = 0x42;
const PING: u8 = 0x43;
const PONG: u8 = 0x44;

pub const ACK_BIT: u8 = 0x08;
pub const SEQ_BIT: u8 = 0x10;
pub const SYNC_MASK: u8 = ACK_BIT | SEQ_BIT;

#[derive(Clone, Copy)]
pub struct ControlByte(u8);

impl ControlByte {
    pub const fn codec_v1() -> Self {
        Self(CODEC_V1)
    }

    pub const fn continuation() -> Self {
        Self(CONTINUATION_PACKET)
    }

    pub const fn ack() -> Self {
        Self(ACK_MESSAGE)
    }

    pub const fn encrypted_transport() -> Self {
        Self(ENCRYPTED_TRANSPORT)
    }

    pub const fn ping() -> Self {
        Self(PING)
    }

    pub const fn pong() -> Self {
        Self(PONG)
    }

    pub const fn error() -> Self {
        Self(ERROR)
    }

    pub const fn channel_allocation_request() -> Self {
        Self(CHANNEL_ALLOCATION_REQ)
    }

    pub const fn channel_allocation_response() -> Self {
        Self(CHANNEL_ALLOCATION_RES)
    }

    pub const fn handshake(phase: HandshakeMessage) -> Self {
        match phase {
            HandshakeMessage::InitiationRequest => Self(HANDSHAKE_INIT_REQ),
            HandshakeMessage::InitiationResponse => Self(HANDSHAKE_INIT_RES),
            HandshakeMessage::CompletionRequest => Self(HANDSHAKE_COMP_REQ),
            HandshakeMessage::CompletionResponse => Self(HANDSHAKE_COMP_RES),
        }
    }

    pub const fn is_ack(&self) -> bool {
        self.0 & ACK_MASK == ACK_MESSAGE
    }

    pub const fn is_continuation(&self) -> bool {
        self.0 & CONTINUATION_PACKET_MASK == CONTINUATION_PACKET
    }

    pub const fn is_encrypted_transport(&self) -> bool {
        self.0 & DATA_MASK == ENCRYPTED_TRANSPORT
    }

    pub const fn is_codec_v1(&self) -> bool {
        self.0 == CODEC_V1
    }

    pub const fn is_ping(&self) -> bool {
        self.0 == PING
    }

    pub const fn is_pong(&self) -> bool {
        self.0 == PONG
    }

    pub const fn is_error(&self) -> bool {
        self.0 == ERROR
    }

    pub const fn is_channel_allocation_request(&self) -> bool {
        self.0 == CHANNEL_ALLOCATION_REQ
    }

    pub const fn is_channel_allocation_response(&self) -> bool {
        self.0 == CHANNEL_ALLOCATION_RES
    }

    pub const fn is_handshake(&self) -> bool {
        self.handshake_phase().is_some()
    }

    pub const fn handshake_phase(&self) -> Option<HandshakeMessage> {
        let res = match self.0 & DATA_MASK {
            HANDSHAKE_INIT_REQ => HandshakeMessage::InitiationRequest,
            HANDSHAKE_INIT_RES => HandshakeMessage::InitiationResponse,
            HANDSHAKE_COMP_REQ => HandshakeMessage::CompletionRequest,
            HANDSHAKE_COMP_RES => HandshakeMessage::CompletionResponse,
            _ => return None,
        };
        Some(res)
    }

    pub fn sync_bits(&self) -> SyncBits {
        SyncBits::from(self.0)
    }

    pub fn with_sync_bits(self, sb: SyncBits) -> Self {
        Self(self.0 & !SYNC_MASK | <SyncBits as Into<u8>>::into(sb))
    }
}

impl TryFrom<u8> for ControlByte {
    type Error = Error;

    fn try_from(byte: u8) -> Result<Self, Error> {
        let cb = Self(byte);
        let valid = cb.is_continuation()
            || cb.is_ack()
            || cb.is_encrypted_transport()
            || cb.is_error()
            || cb.is_ping()
            || cb.is_pong()
            || cb.is_channel_allocation_request()
            || cb.is_channel_allocation_response()
            || cb.is_handshake()
            || cb.is_codec_v1();
        if !valid {
            log::warn!("Invalid control byte {}.", byte);
            return Err(Error::malformed_data());
        }
        Ok(cb)
    }
}

impl From<ControlByte> for u8 {
    fn from(cb: ControlByte) -> Self {
        cb.0
    }
}
